merlin/models/tf/predictions/sampling/in_batch.py

```diff
@@ -0,0 +1,89 @@
+#
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from typing import Optional
+
+import tensorflow as tf
+
+from merlin.models.tf.predictions.sampling.base import Items, ItemSampler
+
+
+@ItemSampler.registry.register("in-batch")
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class InBatchSampler(ItemSampler):
```

Comment:
What is different about this compared to the `InBatchSampler` that exists in `tf.blocks.sampling.in_batch` ?

Reply:
The objective is to move the sampling under the predictions folder as this is used by the new refactored prediction blocks. There are also some enhancements of the sampling blocks (base and InBatchSampler) to improve the API, which are: 

-  Add a registry for sampling strategies 
-  The class arguments are _last_batch_items_embeddings and _last_batch_items_metadata are wrapped in  a new NamedTuple class `Items`.  This new class will replace replace `EmbeddingWithMetadata` class. 
- Add `get_config` method to `InBatchSampler` for serialization. 

I think we could keep the old sampling strategies until we update the High-level API with the new prediction blocks and the old ItemRetrievalTask / ItemRetrievalScorer gets dropped.

Reply:
We can follow-up on this in another PR. I'm wondering if the behavior of this class is the same from an interface/api perspective - will we be able to replace uses of the old one without any changes to the surrounding code.

Reply:
Thanks, that's a good point! Once all tasks defined in this [ticket](https://github.com/NVIDIA-Merlin/Merlin/issues/464) are done, the next step is to do the benchmarking of the retrieval/ranking models and then we change the high-level API. The main constraint when implementing these tasks is to keep the user-facing API the same. 

But in order to simplify things, the item retrieval model will be converted into a Top-k recommender and Top-k evaluation will be implemented differently. The overall objective is to build a `ItemRecommenderModel` class that supports the base methods `predict()` for evaluation and `.save()`  for serialization. [Here](https://github.com/NVIDIA-Merlin/models/compare/main...tf/retrieval-models) is a draft code that summarizes what Marc and I discussed.